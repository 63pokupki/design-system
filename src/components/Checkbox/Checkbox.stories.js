import Checkbox from "./Checkbox.vue";
import { action } from "@storybook/addon-actions";

export default {
    title: "Компоненты/Checkbox",
    component: Checkbox
};

export const empty = () => ({
    components: { Checkbox },
    data() {
        return {
            checked: false
        }
    },
    template: "<Checkbox v-model='checked'></Checkbox>"
});

export const withTextLeft = () => ({
    components: { Checkbox },
    data() {
        return {
            checked: true
        }
    },
    template: `<Checkbox v-model='checked' :position="'left'">Со значком слева</Checkbox>`
});

export const withTextRight = () => ({
    components: { Checkbox },
    data() {
        return {
            checked: true
        }
    },
    template: `<Checkbox v-model='checked' position="right">Со значком справа</Checkbox>`
});

export const primary = () => ({
    components: { Checkbox },
    data() {
        return {
            checked: true
        }
    },
    template: `<Checkbox v-model='checked' type='primary'>Акцентный при нажатии</Checkbox>`
});

export const rounded = () => ({
    components: { Checkbox },
    data() {
        return {
            checked: true
        }
    },
    template: `<Checkbox v-model='checked' :rounded="true" type='primary'>Скругленный</Checkbox>`
});

export const disabled = () => ({
    components: { Checkbox },
    data() {
        return {
            checked: true
        }
    },
    template: `<Checkbox v-model='checked' :disabled="true">Не интерактивный</Checkbox>`
});
