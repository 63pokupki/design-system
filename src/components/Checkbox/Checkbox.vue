<template>
    <label
        :for="uniqid"
        class="spui-Checkbox"
        :class="[_position, _checked, _waiting, _disabled, _rounded, _type]"
    >
        <input
            :checked="value"
            @change="$event => onChange($event.target.checked)"
            class="spui-Checkbox__input"
            type="checkbox"
            :id="uniqid"
        />
        <span v-if="position == 'right'" class="spui-Checkbox__text"><slot></slot></span>
        <span class="spui-Checkbox__fake">
            <i v-if="value" class="spui-Checkbox__icon ds-icon icon-check-in"></i>
        </span>
        <span v-if="position == 'left'" class="spui-Checkbox__text"><slot></slot></span>
    </label>
</template>

<script>
import uuid from "short-uuid";
export default {
    name: "Checkbox",
    props: {
        value: {
            type: Boolean,
            default: false
        },
        name: {
            type: String,
            default: ""
        },
        uniqid: {
            type: String,
            default: uuid.generate()
        },
        disabled: {
            type: Boolean,
            default: false
        },
        type: {
            type: String,
            default: "outline"
        },
        rounded: {
            type: Boolean,
            default: false
        },
        position: {
            type: String,
            default: "left"
        }
    },
    data() {
        return {
            base: "spui-Checkbox"
        };
    },
    computed: {
        _position() {
            return `${this.base}_${this.position}`;
        },
        _checked() {
            return this.value ? `${this.base}_checked` : null;
        },
        _waiting() {
            return !this.value && !this.disabled ? `${this.base}_waiting` : null;
        },
        _disabled() {
            return this.disabled ? `${this.base}_disabled` : null;
        },
        _rounded() {
            return this.rounded ? `${this.base}_rounded` : null;
        },
        _type() {
            return `${this.base}_${this.type}`;
        }
    },
    methods: {
        onChange(value) {
            if (!this.disabled) {
                this.$emit("input", value);
            }
        }
    }
};
</script>

<style lang="scss" scoped>
@import "./Checkbox.scss";
</style>
