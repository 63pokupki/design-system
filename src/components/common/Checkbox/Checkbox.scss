$spui-Checkbox_font: $font-primary;
$spui-Checkbox_colorLabel: #1d1e1f;
$spui-Checkbox_colorActive: #36a6f2;
$spui-Checkbox_colorHover: #0080f6;
$spui-Checkbox_colorDisabled: #c4c4c4;
$spui-Checkbox_borderWidth: 1px;
$spui-Checkbox_borderRadius: 2px;
$spui-Checkbox_width: 16px - $spui-Checkbox_borderWidth * 2;
$spui-Checkbox_height: 17px - $spui-Checkbox_borderWidth * 2;
$spui-Checkbox_fontSize: 12px;

.spui-Checkbox {
    cursor: pointer;
    user-select: none;
    display: inline-flex;

    &__input {
        display: none;
    }
    &__fake {
        position: relative;
        width: $spui-Checkbox_width;
        height: $spui-Checkbox_height;
        min-width: $spui-Checkbox_width;
        min-height: $spui-Checkbox_height;
        border-radius: $spui-Checkbox_borderRadius;
        border: $spui-Checkbox_borderWidth solid $spui-Checkbox_colorActive;
    }
    &__text {
        font-weight: normal;
        font-size: $spui-Checkbox_fontSize;
        font-family: $spui-Checkbox_font;
        color: $spui-Checkbox_colorLabel;
    }

    &__icon {
        color: $spui-Checkbox_colorActive;
        font-size: 10px;
        position: absolute;
        left: 50%;
        top: 50%;
        transform: translate(-50%, -50%);
    }
}

.spui-Checkbox_primary.spui-Checkbox_checked {
    .spui-Checkbox__fake {
        background-color: $spui-Checkbox_colorActive;
    }

    .spui-Checkbox__icon {
        color: #ffffff;
    }
}

.spui-Checkbox_rounded {
    .spui-Checkbox__fake {
        border-radius: 50%;
    }

    .spui-Checkbox__icon {
        padding-top: 1px;
        font-size: 9px;
    }
}

.spui-Checkbox_waiting {
    &:hover {
        .spui-Checkbox__fake {
            border: $spui-Checkbox_borderWidth solid $spui-Checkbox_colorHover;
        }
    }
}

.spui-Checkbox_disabled {
    cursor: default;
    
    .spui-Checkbox__fake {
        border: $spui-Checkbox_borderWidth solid $spui-Checkbox_colorDisabled;
    }

    .spui-Checkbox__text {
        color: $spui-Checkbox_colorDisabled;
    }

    .spui-Checkbox__icon {
        color: $spui-Checkbox_colorDisabled;
    }
}

.spui-Checkbox_left {
    .spui-Checkbox__text {
        padding-left: 8px;
    }
}

.spui-Checkbox_right {
    .spui-Checkbox__text {
        padding-right: 8px;
    }
}

.spui-Checkbox_only-text {
    .spui-Checkbox__fake {
        display: none;
    }
    .spui-Checkbox__text {
        padding: 0;
    }
}

.spui-Checkbox_align-center {
    align-items: center;
}
