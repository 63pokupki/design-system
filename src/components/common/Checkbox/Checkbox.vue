<template>
    <label
        class="spui-Checkbox"
        :class="[
            _position,
            _checkedClass,
            _waiting,
            _disabled,
            _rounded,
            _type,
            _onlyText,
            _align,
        ]"
    >
        <input
            v-model="_model"
            :value="val"
            class="spui-Checkbox__input"
            type="checkbox"
        >
        <span
            v-if="position == 'right' && $slots.default"
            class="spui-Checkbox__text"
        >
            <slot name="default" />
        </span>
        <div class="spui-Checkbox__fake">
            <i
                v-if="_checked"
                class="spui-Checkbox__icon ds-icon icon-check-in-checkbox"
            />
        </div>
        <span
            v-if="position == 'left' && $slots.default"
            class="spui-Checkbox__text"
        >
            <slot name="default" />
        </span>
    </label>
</template>

<script>
import isEqual from 'lodash-es/isEqual';

export default {
    name: 'Checkbox',
    props: {
        /** Модель */
        value: {
            type: [Array, Boolean],
            default: false,
        },
        /** Значение */
        val: {
            type: [Array, Object, String, Boolean, Number],
            default: 'Значение не передано',
        },
        /** Отображать только текст без галочки */
        onlyText: {
            type: Boolean,
            default: false,
        },
        /** Вертикальное центрирование галочки и текста */
        alignCenter: {
            type: Boolean,
            default: true,
        },
        /** Неактивное состояние */
        disabled: {
            type: Boolean,
            default: false,
        },
        /** Внешний вид */
        type: {
            type: String,
            default: 'outline',
            validator(value) {
                return ['outline', 'primary'].includes(value);
            },
        },
        /** Круглый вид */
        rounded: {
            type: Boolean,
            default: false,
        },
        /** Позиция галочки относительно текста */
        position: {
            type: String,
            default: 'left',
            validator(value) {
                return ['left', 'right'].includes(value);
            },
        },
    },
    data() {
        return {
            base: 'spui-Checkbox',
        };
    },
    computed: {
        /** Модель */
        _model: {
            get() {
                return this.value;
            },
            set(value) {
                if (!this.disabled) {
                    this.$emit('input', value);
                }
            },
        },
        /** Стилевой класс показа только текста */
        _onlyText() {
            return this.onlyText && `${this.base}_only-text`;
        },
        /** Стилевой класс позиции текста относительно галочки */
        _position() {
            return `${this.base}_${this.position}`;
        },
        /** Состояние - выбран или нет */
        _checked() {
            const model = this._model;
            const value = this.val;

            if (Array.isArray(model)) {
                return model.some((val) => isEqual(val, value));
            }
            return model;
        },
        /** Стилевой класс состояния - выбран или нет */
        _checkedClass() {
            return this._checked && `${this.base}_checked`;
        },
        /** Стилевой класс вертикального центрирования галочки и текста */
        _align() {
            return this.alignCenter && `${this.base}_align-center`;
        },
        /** Стилевой класс состояния - не выбран, не отключен */
        _waiting() {
            return !this._checked && !this.disabled && `${this.base}_waiting`;
        },
        /** Стилевой класс состояния отключен */
        _disabled() {
            return this.disabled && `${this.base}_disabled`;
        },
        /** Стилевой класс круглого вида */
        _rounded() {
            return this.rounded && `${this.base}_rounded`;
        },
        /** Стилевой класс вида */
        _type() {
            return `${this.base}_${this.type}`;
        },
    },
};
</script>

<style lang="scss" scoped>
@import './Checkbox.scss';
</style>
