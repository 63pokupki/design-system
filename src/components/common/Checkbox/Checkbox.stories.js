import Checkbox from "./Checkbox.vue";

export default {
    title: "Общие/Checkbox",
    component: Checkbox
};

export const modelBoolean = () => ({
    components: { Checkbox },
    data() {
        return {
            checked: false
        }
    },
    template: `
    <div>
        <p>Значение: {{checked}}</p>
        <Checkbox v-model='checked'></Checkbox>
    </div>
    `
});

export const modelArray = () => ({
    components: { Checkbox },
    data() {
        return {
            checked: []
        }
    },
    template: `
    <div>
        <p>Значение: {{checked}}</p>
        <Checkbox val="Значение 1" v-model='checked'>Значение 1</Checkbox>
        <Checkbox val="Значение 2" v-model='checked'>Значение 2</Checkbox>
        <Checkbox val="Значение 3" v-model='checked'>Значение 3</Checkbox>
    </div>
    `
});

export const onlyText = () => ({
    components: { Checkbox },
    data() {
        return {
            checked: []
        }
    },
    template: `
    <div>
        <p>Значение: {{checked}}</p>
        <Checkbox val="Значение 1" :onlyText="true" v-model='checked'>Значение 1</Checkbox>
        <Checkbox val="Значение 2" :onlyText="true" v-model='checked'>Значение 2</Checkbox>
        <Checkbox val="Значение 3" :onlyText="true" v-model='checked'>Значение 3</Checkbox>
    </div>
    `
});

export const withTextLeft = () => ({
    components: { Checkbox },
    data() {
        return {
            checked: true
        }
    },
    template: `<Checkbox v-model='checked' position="left">Со значком слева</Checkbox>`
});

export const withTextRight = () => ({
    components: { Checkbox },
    data() {
        return {
            checked: true
        }
    },
    template: `<Checkbox v-model='checked' position="right">Со значком справа</Checkbox>`
});

export const primary = () => ({
    components: { Checkbox },
    data() {
        return {
            checked: true
        }
    },
    template: `<Checkbox v-model='checked' type='primary'>Акцентный при нажатии</Checkbox>`
});

export const rounded = () => ({
    components: { Checkbox },
    data() {
        return {
            checked: true
        }
    },
    template: `<Checkbox v-model='checked' :rounded="true" type='primary'>Скругленный</Checkbox>`
});

export const disabled = () => ({
    components: { Checkbox },
    data() {
        return {
            checked: true
        }
    },
    template: `<Checkbox v-model='checked' :disabled="true">Не интерактивный</Checkbox>`
});
