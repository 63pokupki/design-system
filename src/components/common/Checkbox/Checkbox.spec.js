import { mount, createLocalVue } from '@vue/test-utils';
import Checkbox from '@/components/common/Checkbox/Checkbox.vue';

/** Фабрика создания компонента */
const mountComponent = (context = {}) => {
    const localVue = createLocalVue();
    return mount(Checkbox, {
        ...context,
        localVue,
    });
};

describe('Checkbox.vue', () => {
    it('Инициализация', () => {
        const wrapper = mountComponent();
        expect(wrapper.exists()).toBeTruthy();
    });

    it('Проверка стандартного состояния', () => {
        const wrapper = mountComponent({
            propsData: {
                id: 'id',
            },
        });

        expect(wrapper.find('input[type=checkbox]').element.checked).toBe(
            false,
        );
        expect(wrapper.vm.$data.uniqid).toBe('id');
        expect(wrapper.find('.spui-Checkbox__text').element).toBeUndefined();

        expect(wrapper.html()).toMatchSnapshot();
    });
});
