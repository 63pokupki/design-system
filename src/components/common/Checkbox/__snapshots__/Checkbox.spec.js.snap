// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox.vue Проверка стандартного состояния 1`] = `
<label for="id" class="spui-Checkbox spui-Checkbox_left spui-Checkbox_waiting spui-Checkbox_outline spui-Checkbox_align-center"><input id="id" type="checkbox" class="spui-Checkbox__input" value="Значение не передано">
  <!---->
  <div class="spui-Checkbox__fake">
    <!---->
  </div>
  <!---->
</label>
`;
